### PR TITLE
Use the LLVM New Pass Manager

### PR DESCRIPTION
**Pull request by lmontigny**
(rebased from origin/main, #334 to delete)

[o] Need to add function test
[o] Solve Aggregation overflow issue
[x] Need to check performance of tests (was using -O3 opt by default)

Current performance:
```
[==========] Running 8 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 8 tests from CodeGeneratorTest
[ RUN      ] CodeGeneratorTest.IntegerConstant
[       OK ] CodeGeneratorTest.IntegerConstant (186 ms)
[ RUN      ] CodeGeneratorTest.IntegerAdd
[       OK ] CodeGeneratorTest.IntegerAdd (173 ms)
[ RUN      ] CodeGeneratorTest.IntegerColumn
[       OK ] CodeGeneratorTest.IntegerColumn (174 ms)
[ RUN      ] CodeGeneratorTest.IntegerExpr
[       OK ] CodeGeneratorTest.IntegerExpr (174 ms)
[ RUN      ] CodeGeneratorTest.IntegerConstantGPU
[       OK ] CodeGeneratorTest.IntegerConstantGPU (290 ms)
[ RUN      ] CodeGeneratorTest.IntegerAddGPU
[       OK ] CodeGeneratorTest.IntegerAddGPU (265 ms)
[ RUN      ] CodeGeneratorTest.IntegerColumnGPU
[       OK ] CodeGeneratorTest.IntegerColumnGPU (294 ms)
[ RUN      ] CodeGeneratorTest.IntegerExprGPU
[       OK ] CodeGeneratorTest.IntegerExprGPU (266 ms)
[----------] 8 tests from CodeGeneratorTest (1822 ms total)

[----------] Global test environment tear-down
[==========] 8 tests from 1 test suite ran. (1822 ms total)
[  PASSED  ] 8 tests.

```

### CODE REVIEW

File: omniscidb/QueryEngine/Compiler/HelperFunctions.cpp [hunk 1]
```diff
@@ -89,47 +89,66 @@ void eliminate_dead_self_recursive_funcs(
...
+  llvm::CGSCCPassManager CGPM;
+
+  PB.registerModuleAnalyses(MAM);
+  PB.registerCGSCCAnalyses(CGAM);
+  PB.registerFunctionAnalyses(FAM);
+  PB.registerLoopAnalyses(LAM);
+  PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
+  llvm::ModulePassManager MPM =
+      PB.buildPerModuleDefaultPipeline(llvm::OptimizationLevel::O3);
```
Comment (kurapov-peter) [inline omniscidb/QueryEngine/Compiler/HelperFunctions.cpp:112]: This adds a bunch of passes that are likely not needed and most likely cost us a lot of compile time.

File: omniscidb/QueryEngine/Compiler/HelperFunctions.h [hunk 2]
```diff
@@ -17,9 +17,25 @@
 #include <string>
 #include <unordered_set>
 
+#include <llvm/Analysis/MemorySSA.h>
 #include <llvm/IR/LegacyPassManager.h>
```
Comment (kurapov-peter) [inline omniscidb/QueryEngine/Compiler/HelperFunctions.h:21]: You can get rid of the legacy one.

File: omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 #include <llvm/Analysis/CallGraph.h>
 #include <llvm/Analysis/CallGraphSCCPass.h>
```
Comment (kurapov-peter) [inline omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h:20]: Same here.

File: omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h [hunk 4]
```diff
@@ -64,11 +76,9 @@ class AnnotateInternalFunctionsPass : public llvm::CallGraphSCCPass {
       }
     }
 
-    return updated_function_defs;
+    return llvm::PreservedAnalyses::all();
```
Comment (kurapov-peter) [inline omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h:79]: The change probably breaks some analyses, we should double-check it.

**Review by kurapov-peter — CHANGES_REQUESTED** [review]

File: omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h (hunk 4 above)
Comment (lmontigny) [reply omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h:79]: With the new PM, I need to return PA or PreservedAnalyses::none(); or llvm::PreservedAnalyses::all(); for example

Comment (kurapov-peter) [reply omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h:79]: `llvm::PreservedAnalyses` represents a set of analyses that are preserved by a transformation pass. You can create such an object and specify which analyses are preserved/abandoned. The `none()` and `all()` are just convenience methods.

File: omniscidb/QueryEngine/Compiler/HelperFunctions.cpp (hunk 1 above)
Comment (lmontigny) [reply omniscidb/QueryEngine/Compiler/HelperFunctions.cpp:112]:
```
lmontign@ansatlin08:/localdisk/lmontign/hdk/omniscidb/build$ ./Tests/CodeGeneratorTest
[==========] Running 8 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 8 tests from CodeGeneratorTest
[ RUN      ] CodeGeneratorTest.IntegerConstant
[       OK ] CodeGeneratorTest.IntegerConstant (186 ms)
[ RUN      ] CodeGeneratorTest.IntegerAdd
[       OK ] CodeGeneratorTest.IntegerAdd (176 ms)
[ RUN      ] CodeGeneratorTest.IntegerColumn
[       OK ] CodeGeneratorTest.IntegerColumn (175 ms)
[ RUN      ] CodeGeneratorTest.IntegerExpr
[       OK ] CodeGeneratorTest.IntegerExpr (177 ms)
[ RUN      ] CodeGeneratorTest.IntegerConstantGPU
[       OK ] CodeGeneratorTest.IntegerConstantGPU (283 ms)
[ RUN      ] CodeGeneratorTest.IntegerAddGPU
[       OK ] CodeGeneratorTest.IntegerAddGPU (576 ms)
[ RUN      ] CodeGeneratorTest.IntegerColumnGPU
[       OK ] CodeGeneratorTest.IntegerColumnGPU (289 ms)
[ RUN      ] CodeGeneratorTest.IntegerExprGPU
[       OK ] CodeGeneratorTest.IntegerExprGPU (314 ms)
[----------] 8 tests from CodeGeneratorTest (2177 ms total)

[----------] Global test environment tear-down
[==========] 8 tests from 1 test suite ran. (2177 ms total)
[  PASSED  ] 8 tests.

```

File: omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h (hunk 4 above)
Comment (lmontigny) [reply omniscidb/QueryEngine/Optimization/AnnotateInternalFunctionsPass.h:79]: I see. We would need to check how to test if the analyses are broken or not.

File: omniscidb/QueryEngine/Compiler/Backend.cpp [hunk 5]
```diff
@@ -84,9 +84,9 @@ std::shared_ptr<CpuCompilationContext> CPUBackend::generateNativeCPUCode(
   llvm::Module* llvm_module = func->getParent();
   // run optimizations
 #ifndef WITH_JIT_DEBUG
-  llvm::legacy::PassManager pass_manager;
+  llvm::PassBuilder pass_builder;
   compiler::optimize_ir(
-      func, llvm_module, pass_manager, live_funcs, /*is_gpu_smem_used=*/false, co);
+      func, llvm_module, pass_builder, live_funcs, /*is_gpu_smem_used=*/false, co);
```
Comment (kurapov-peter) [inline omniscidb/QueryEngine/Compiler/Backend.cpp:89]: We can probably get rid of the builder parameter, I don't think we prepopulate it anywhere.

**Comment by lmontigny** [comment]
Overflow issue for the **AggregationTest.FilterAndSum** test with `./Tests/IntelGPUEnablingTest  --`

```
[ RUN      ] AggregationTest.FilterAndSum
2023-04-03T07:13:07.109105 E 2823166 0 0 RelAlgExecutor.cpp:1346 Query execution failed with error ERR_OVERFLOW_OR_UNDERFLOW: Overflow or underflow
unknown file: Failure
C++ exception with description "ERR_OVERFLOW_OR_UNDERFLOW: Overflow or underflow" thrown in the test body.
[  FAILED  ] AggregationTest.FilterAndSum (989 ms)
```

**Comment by kurapov-peter** [comment]
We can do better with the multiplication overflow check: https://godbolt.org/z/zGf53jq5q

**Comment by lmontigny** [comment]
After comparison with the HEAD, the `mul_check.i` was different in the IR.

The overflow comes from this test:
```
TEST_F(AggregationTest, FilterAndSum) {
  ...
  c("SELECT SUM(dd * ufd) FROM test;", g_dt);
```

A potential reason would be the overflow check algo used in:
llvm::Value* CodeGenerator::codegenMul

It can be simplified

**Comment by lmontigny** [comment]
With the GNU overflow check, there is still an overflow with the following case:
`((b < 0) && (a > 0) && (a > std::numeric_limits<T>::min() / b));`

Need more details.

```
Input test: dd, ufd
Column types:
DEC64(10,2)|INT32[NN]
Data:
11110|null
11110|null
11110|null
11110|null
11110|null
11110|null
11110|null
11110|null
11110|null
11110|null
22220|-2147483647
22220|-2147483647
22220|-2147483647
22220|-2147483647
22220|-2147483647
33330|-1
33330|-1
33330|-1
33330|-1
33330|-1

```

**Comment by lmontigny** [comment]
Implemented the [GNU overflow check ](https://github.com/coreutils/gnulib/blob/master/lib/intprops.h#L135)in `codegenMul` , it seems to work correctly.
However it affected the performance as it created 5 control flow branches.

Considering using the [LLVM check ](https://github.com/hdoc/llvm-project/blob/release/15.x//llvm/include/llvm/Support/MathExtras.h#L976) instead.

**Review by kurapov-peter — CHANGES_REQUESTED** [review]
Could you rebase and clean it up please?

**Comment by lmontigny** [comment]
> Could you rebase and clean it up please?

Switched to https://github.com/intel-ai/hdk/pull/417

**Pull request closed without merging**